### DISCUSSION TITLE
Fix curtain control module state "unknown"

### DISCUSSION THREAD

**Discussion by naftalih** [make-all/tuya-local#1681] [Show and tell]
On cover.py _current_state function, "stop" command is not handled for devices with _control_dp, and therefore when curtain get the stop command it's state is not updated.

Fix-

Edit cover.py under tuya_local custom components and add the following lines at the end of the _current_state function under self._control_dp if statement (after line 154):

elif cmd == "stop":
      if pos == 0:
          return "closed"
      else:
          return "opened"

I hope the developers will include this in future releases.

![20240225_003954](https://github.com/make-all/tuya-local/assets/48178152/8d3d73d8-3976-45cb-981e-596d505d91fb)

**Comment by make-all** [comment]
This is incorrect.  If a device has a dp that reports the current position of the curtain, it is handled in the block above that. The changes you have made are based on the last position the user requested the curtain to move to, which is often not updated at all, and may have been several  commands ago.  The only thing you can determine from such a curtain device is that the state is unknown.

**Reply by naftalih** [reply]
Hi, my module doesn't have current position dp. Therefore it works only based on the command state + percentage position.

**Reply by naftalih** [reply]
BTW- after my changes state is perfect.

**Reply by naftalih** [reply]
My module have the following dps:
1 control (open, stop, close)
2 position control (percentage)
8 motor direction
10 calibration

**Reply by naftalih** [reply]
You can get current position with dp 2 but your current_position is using specific dp which is not 2.

**Reply by naftalih** [reply]
Maybe I should mention that this module's cmd state is always "stop" when motor is not moving.

**Comment by make-all** [comment]
This is intended to be handled by listing dp 2 as both position and current_position in the config.  I have added `curtains_with_feedback.yaml` based on the qs_c01_curtain.yaml that matches those dps above with the additional current_position mapping.

It may apply to all QS C01 curtains, but my experience with Tuya curtains is that there are many variations in behaviour with the same dps mapping, so I thought it will be better to make a separate config.

**Comment by naftalih** [comment]
Hi Jason,

This device (NH-ZS 108) doesn't have a specific dp for "current position".
It got only 5 dps:
1. control (open, stop, close)
2. percent_control (percantage)
3. cur_calibration (start, end)
8. Motor direction (control_back_mode)
10. Calibration time (tr_timecon)

You get "current position" from dp 2.

This is a module for push button switch. Meaning that the stop command is
handled by the module in order to stop operation (either at a certain
percentage or at the end of opening or closing by duration time calibrated).

So the only way to determine "is_closed" is by checking the combination of
0 percent (100 percent in case of inversion) position and cmd is "stop".

Thanks,
Naftali
